balancer/pickfirst/pickfirstleaf/pickfirstleaf_ext_test.go

```diff
@@ -851,6 +850,245 @@ func (s) TestPickFirstLeaf_EmptyAddressList(t *testing.T) {
 	}
 }
 
+// TestPickFirstLeaf_HappyEyeballs_TFAfterEndOfList verifies that pickfirst
+// correctly detects the end of the first happy eyeballs pass when the timer
+// causes pickfirst to reach the end of the address list and failures are
+// reported out of order.
+func (s) TestPickFirstLeaf_HappyEyeballs_TFAfterEndOfList(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	timerCh := make(chan struct{})
+	originalTimer := pfinternal.TimeAfterFunc
+	pfinternal.TimeAfterFunc = func(_ time.Duration, f func()) *time.Timer {
+		// Set a really long expiration to prevent it from triggering
+		// automatically.
+		timer := time.AfterFunc(time.Hour, f)
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-timerCh:
+			}
+			timer.Reset(0)
+		}()
+		return timer
+	}
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	dialer := testutils.NewBlockingDialer()
+	opts := []grpc.DialOption{
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, pickfirstleaf.Name)),
+		grpc.WithContextDialer(dialer.DialContext),
+	}
+	cc, rb, bm := setupPickFirstLeaf(t, 3, opts...)
+	addrs := bm.resolverAddrs()
+	holds := bm.holds(dialer)
+	rb.UpdateState(resolver.State{Addresses: addrs})
+	cc.Connect()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Verify that only the first server is contacted.
+	if holds[0].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 0, addrs[0])
+	}
+
+	// Ensure no other servers are contacted.
+	if holds[1].IsStarted() != false {
+		t.Fatalf("Server %d with address %q contacted unexpectedly", 1, addrs[1])
+	}
+	if holds[2].IsStarted() != false {
+		t.Fatalf("Server %d with address %q contacted unexpectedly", 2, addrs[2])
+	}
+
+	// Make the happy eyeballs timer fire once and verify that the
+	// second server is contacted, but the third isn't.
+	timerCh <- struct{}{}
+
+	// Verify that the second server is contacted and 3rd isn't.
```

Comment:
Nit: You could remove the newline and this comment since the comment on line 907 conveys the same thing and the two verifications that are below are covered as part of that comment. Thanks.